### ISSUE TITLE
Scheduled Jobs not Terminating

### ISSUE THREAD

**Issue by ameissnersofi** [rundeck/rundeck#2235]
# Bug report

**My Rundeck detail**

* Rundeck version: 2.6.11
* install type: launcher
* OS Name/version: alpine linux 3.4

**Expected Behavior**

Scheduled jobs should run and terminate with appropriate status.

**Actual Behavior**

Scheduled jobs sometimes enter a state where they have finished running (at least according to log output) but the rundeck execution never officially terminates.  This prevents the same job from being run again.  I have to login to rundeck and kill the job.  This, in turn, disables the schedule, so I have to go and re-enable that as well.

The big problem is the jobs never seem to terminate.

**How to reproduce Behavior**

The jobs in question have one step to them that is executing a python script (logic in rundeck).  This python script makes several remote API calls to other machines, which all complete successfully.  Again, I see my "Done" log message at the bottom of the Log Output pane, but the job never quits.

**Comment by ameissnersofi** [comment]
Here's a screen shot of the running pane after i killed the jobs.  These should never run for that long.  I had to kill the first listed job before sending this along

![screen shot 2016-12-09 at 2 13 44 pm](https://cloud.githubusercontent.com/assets/22750865/21064741/cd5752e6-be19-11e6-86a8-a5ffc2a46c33.png)

**Comment by ameissnersofi** [comment]
After killing one of the stalled jobs, I got this stacktrace:

```
rundeck.services.ExecutionServiceException: Job "Consul_Monitor" {{Job ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d}} is currently being executed {{Execution 604970}}
        at rundeck.services.ExecutionService.createExecution(ExecutionService.groovy:1965)
        at rundeck.quartzjobs.ExecutionJob.initialize(ExecutionJob.groovy:291)
        at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:111)
        at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:90)
        at com.codahale.metrics.Timer.time(Timer.java:99)
        at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:89)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
2016-12-10 00:28:00,157 - [] [] [/] [#] [] - Job Monitors:Consul_Monitor:.19341:Consul_Monitor threw an unhandled Exception:
rundeck.services.ExecutionServiceException: Job "Consul_Monitor" {{Job ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d}} is currently being executed {{Execution 604970}}
        at rundeck.services.ExecutionService.createExecution(ExecutionService.groovy:1965)
        at rundeck.quartzjobs.ExecutionJob.initialize(ExecutionJob.groovy:291)
        at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:111)
        at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:90)
        at com.codahale.metrics.Timer.time(Timer.java:99)
        at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:89)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
2016-12-10 00:28:00,159 - [] [] [/] [#] [] - Job (Monitors:Consul_Monitor:.19341:Consul_Monitor threw an exception.
org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: rundeck.services.ExecutionServiceException: Job "Consul_Monitor" {{Job ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d}} is currently being executed {{Execution 604970}}]
        at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: rundeck.services.ExecutionServiceException: Job "Consul_Monitor" {{Job ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d}} is currently being executed {{Execution 604970}}
        at rundeck.services.ExecutionService.createExecution(ExecutionService.groovy:1965)
        at rundeck.quartzjobs.ExecutionJob.initialize(ExecutionJob.groovy:291)
        at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:111)
        at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:90)
        at com.codahale.metrics.Timer.time(Timer.java:99)
        at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:89)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        ... 1 more
2016-12-10 00:28:00,164 - [] [] [/] [#] [] - Exception occurred in job: Attempts to heal failing nodes before alerting.
org.quartz.JobExecutionException: org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: rundeck.services.ExecutionServiceException: Job "Consul_Monitor" {{Job ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d}} is currently being executed {{Execution 604970}}] [See nested exception: org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: rundeck.services.ExecutionServiceException: Job "Consul_Monitor" {{Job ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d}} is currently being executed {{Execution 604970}}]]
        at org.quartz.core.JobRunShell.run(JobRunShell.java:218)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: rundeck.services.ExecutionServiceException: Job "Consul_Monitor" {{Job ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d}} is currently being executed {{Execution 604970}}]
        at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
        ... 1 more
Caused by: rundeck.services.ExecutionServiceException: Job "Consul_Monitor" {{Job ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d}} is currently being executed {{Execution 604970}}
        at rundeck.services.ExecutionService.createExecution(ExecutionService.groovy:1965)
        at rundeck.quartzjobs.ExecutionJob.initialize(ExecutionJob.groovy:291)
        at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:111)
        at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:90)
        at com.codahale.metrics.Timer.time(Timer.java:99)
        at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:89)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        ... 1 more
2016-12-10 00:28:10,408 - [ameissner] [Monitors] [/Consul_Monitor] [#604970] [ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d] - Result: 2147483647
2016-12-10 00:28:10,817 - [ameissner] [Monitors] [/Consul_Monitor] [#604970] [ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d] - Failed: NonZeroResultCode: Result code was 2147483647
2016-12-10 00:28:10,847 - [] [] [/] [#] [] - Execution failed: 604970 in project Monitors: [Workflow result: , step failures: {1=Dispatch failed on 1 nodes: [internal.sofi.com: NonZeroResultCode: Result code was 2147483647]}, Node failures: {internal.sofi.com=[NonZeroResultCode: Result code was 2147483647]}, status: failed]
2016-12-10 00:28:10,856 - [] [] [/] [#] [] - Failed addEvent for StreamingLogWriter(log4jStreamingLogWriterPlugin), disabling. Ambiguous method overloading for method java.util.HashMap#plus.
Cannot resolve which method to invoke for [null] due to overlapping prototypes between:
        [interface java.util.Collection]
        [interface java.util.Map]
2016-12-10 00:28:10,408 project: Monitors execution: 604970 (log, ) /Consul_Monitor [ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d] - Result: 2147483647
2016-12-10 00:28:10,817 project: Monitors execution: 604970 (log, ) /Consul_Monitor [ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d] - Failed: NonZeroResultCode: Result code was 2147483647
```

**Comment by ameissnersofi** [comment]
I don't mean to be a nuisance, and I know you have lots of other things to work on - but has anybody taken a look at this?  This is kind of mission critical for us as our infrastructure's monitors are housed in rundeck.  We've found a temporary solution in the interim, but we really need rundeck to facilitate this.

Thanks again!

**Comment by gschueler** [comment]
This sounds like some error is occurring at the time the execution ends, preventing it from entering a finished state and rescheduling.  Can you check your service.log for error messages?

**Comment by aristosv** [comment]
same issue here. reported it here
https://groups.google.com/forum/#!topic/rundeck-discuss/I5ow7yENkF8

**Comment by nboury-vsct** [comment]
Hi !
same behavior for us. It seems to appear when we have a long list of target nodes for the execution (for other executions with one or two target nodes, no problem)

**Comment by jamieps** [comment]
It sounds like two issues are being conflated...

@aristosv your issue is #2228 which was introduced in v2.7 and fixed in v2.7.2 (not yet released).
The OP is reporting this on v2.6.11 so shouldn't be affected by #2228.

@nboury-vsct I'm assuming that you have the same errors as reported by @aristosv, in which case it's been fixed already.

**Comment by ameissnersofi** [comment]
@gschueler sorry, i put this on the wrong issue.  It's supposed to be here:

`rundeck.log` (since i did the launher install) is littered with 

```
2016-12-30 10:00:00,188 [quartzScheduler_Worker-7] ERROR org.quartz.core.JobRunShell - Job Monitors:Consul_Monitor:STALLS/PYTHON.19341:Consul_Monitor threw an unhandled Exception: 
rundeck.services.ExecutionServiceException: Job "Consul_Monitor" {{Job ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d}} is currently being executed {{Execution 500227}}
	at rundeck.services.ExecutionService.createExecution(ExecutionService.groovy:1965)
	at rundeck.quartzjobs.ExecutionJob.initialize(ExecutionJob.groovy:291)
	at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:111)
	at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:90)
	at com.codahale.metrics.Timer.time(Timer.java:99)
	at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:89)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
2016-12-30 10:00:00,193 [quartzScheduler_Worker-7] ERROR org.quartz.core.ErrorLogger - Job (Monitors:Consul_Monitor:STALLS/PYTHON.19341:Consul_Monitor threw an exception.
org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: rundeck.services.ExecutionServiceException: Job "Consul_Monitor" {{Job ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d}} is currently being executed {{Execution 500227}}]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: rundeck.services.ExecutionServiceException: Job "Consul_Monitor" {{Job ac0c4097-0f43-494f-ab42-d7fdbb3f8c9d}} is currently being executed {{Execution 500227}}
	at rundeck.services.ExecutionService.createExecution(ExecutionService.groovy:1965)
	at rundeck.quartzjobs.ExecutionJob.initialize(ExecutionJob.groovy:291)
	at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:111)
	at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:90)
	at com.codahale.metrics.Timer.time(Timer.java:99)
	at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:89)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	... 1 more
```

**Comment by zolech** [comment]
Same problem here, except my script is outputting with exit code 1 (fail). Whole job should be marked as failed but instead it is running forever and can't be killed. We have to restart whole rundeck service because job is set not to run with multiple executions.

`2017-01-13 14:00:21,802 [quartzScheduler_Worker-18] ERROR org.quartz.core.JobRunShell - Job "Jobname1" threw an unhandled Exception:   
rundeck.services.ExecutionServiceException: Job "Jobname1" [d3b191ab-7a57-476f-aa11-faa6ce619a5e] is currently being executed (execution [[2198169]])
        at rundeck.services.ExecutionService.createExecution(ExecutionService.groovy:1965)
        at rundeck.quartzjobs.ExecutionJob.initialize(ExecutionJob.groovy:291)
        at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:111)
        at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:90)
        at com.codahale.metrics.Timer.time(Timer.java:99)
        at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:89)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
2017-01-13 14:00:21,802 [quartzScheduler_Worker-18] ERROR org.quartz.core.ErrorLogger - Job (Jobname1 threw an exception.
org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: rundeck.services.ExecutionServiceException: Job "Jobname1" [d3b191ab-7a57-476f-aa11-faa6ce619a5e] is currently being executed (execution [[2198169]])]
        at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: rundeck.services.ExecutionServiceException: Job "Jobname1" [d3b191ab-7a57-476f-aa11-faa6ce619a5e] is currently being executed (execution [[2198169]])
        at rundeck.services.ExecutionService.createExecution(ExecutionService.groovy:1965)
        at rundeck.quartzjobs.ExecutionJob.initialize(ExecutionJob.groovy:291)
        at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:111)
        at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:90)
        at com.codahale.metrics.Timer.time(Timer.java:99)
        at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:89)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        ... 1 more`

**Comment by zolech** [comment]
Small update from my side. At the time when it happens there is always a deadlock on a row corresponding to job in db (we use mysql 5.6). Removing deadlock and manually updating row to mark job as failed works and I don't have to restart whole scheduler. However doing it couple of times a days is not funny. Why is it deadlocking? I always thought only one process is updating a row at the time.

**Comment by gschueler** [comment]
@zolech can you do a thread dump when you get a deadlock like that?

**Comment by zolech** [comment]
Sure. Will dump of "metrics/threads" enpoint be sufficient?

**Comment by gschueler** [comment]
should be, or you can do `kill -QUIT` on the java process which should dump the threads as well

**Comment by zolech** [comment]
Ok I wouldn't have to wait too long. Right now I have one job that returned error -1 and should be marked as failed. Instead it is "Running"

![job](https://cloud.githubusercontent.com/assets/2222322/22752724/e3d8ad94-ee39-11e6-8757-d7a37a2216f1.png)

 Also there is a deadlock in a database:

`---TRANSACTION 2698750092, ACTIVE 5730 sec
3 lock struct(s), heap size 360, 1 row lock(s), undo log entries 2
MySQL thread id 8040113, OS thread handle 0x7fd3e74d4700, query id 350537822 <<client ip>> <<user>> cleaning up
Trx read view will not see trx with id >= 2698750093, sees < 2698749969
`

Here is a thread dump
[dump.txt](https://github.com/rundeck/rundeck/files/761691/dump.txt)

**Comment by zolech** [comment]
Hi, I have a little update on that. 

We've been using slack notification plugin https://github.com/higanworks/rundeck-slack-incoming-webhook-plugin. After disabling notifications for jobs that were constantly "hanging" not even one hanged during last week (it was 4-5 times a day before). So it is either connected with this plugin or with sending notifications itself.

**Comment by gschueler** [comment]
@zolech thanks, that sounds like perhaps an error when running the notification is causing the state to not be saved correctly

**Comment by wangchuande** [comment]
I have the same problem.
Rundeck Version: 2.7.3
rundeck.log

```java
2017-05-15 05:00:00,029 [quartzScheduler_Worker-2] ERROR org.quartz.core.JobRunShell - Job Postgres:restore_roc:.17992:restore_r
oc threw an unhandled Exception:
rundeck.services.ExecutionServiceException: Job "restore_roc" {{Job 2c7a16f4-d191-41f0-bfb5-84dfa3c1b471}} is currently being ex
ecuted {{Execution 42249}}
	at rundeck.services.ExecutionService.createExecution(ExecutionService.groovy:1955)
	at rundeck.quartzjobs.ExecutionJob.initialize(ExecutionJob.groovy:300)
	at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:112)
	at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:91)
	at com.codahale.metrics.Timer.time(Timer.java:99)
	at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:90)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
2017-05-15 05:00:00,030 [quartzScheduler_Worker-2] ERROR org.quartz.core.ErrorLogger - Job (Postgres:restore_roc:.17992:restore_
roc threw an exception.
org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: rundeck.services.ExecutionServiceExcepti
on: Job "restore_roc" {{Job 2c7a16f4-d191-41f0-bfb5-84dfa3c1b471}} is currently being executed {{Execution 42249}}]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: rundeck.services.ExecutionServiceException: Job "restore_roc" {{Job 2c7a16f4-d191-41f0-bfb5-84dfa3c1b471}} is current
ly being executed {{Execution 42249}}
	at rundeck.services.ExecutionService.createExecution(ExecutionService.groovy:1955)
	at rundeck.quartzjobs.ExecutionJob.initialize(ExecutionJob.groovy:300)
	at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:112)
	at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:91)
	at com.codahale.metrics.Timer.time(Timer.java:99)
	at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:90)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	... 1 more
```

**Comment by Maannniii** [comment]
This [closed issue](https://github.com/rundeck/rundeck/issues/2228) is the solution for the problem.

**Comment by acasademont** [comment]
Hi all and @gschueler 
We discovered that a notification system that pings an external source (Eg: Datadog) can timeout and by default, the URL class has an undefined timeout, meaning that the request will enver end and Rundeck will basically never terminate the task. We had 3 tasks stalled for over 3 hours and after quite a lot of digging we found 3 open sockets to the Datadog endpoint that never terminated. We have explicitly set both a connection timeout and a read timeout in our Datadog notification plugin. Will let the thing run for a few days and see what happens.
